File: app/src/main/java/pl/vemu/zsme/detailedNews/galleryFragment/GalleryFragment.java
package pl.vemu.zsme.detailedNews.galleryFragment;

import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowInsets;
import android.view.WindowInsetsController;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.bottomnavigation.BottomNavigationView;

import org.jetbrains.annotations.NotNull;

import pl.vemu.zsme.R;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;

public class GalleryFragment extends Fragment implements SetUIVisibility {

    private ViewPager2 pager;
    private boolean isUiVisible = true;

    public GalleryFragment() {
    }

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        pager = new ViewPager2(getContext());
        pager.setLayoutParams(new ViewGroup.LayoutParams(MATCH_PARENT, MATCH_PARENT));
        return pager;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        String[] images = GalleryFragmentArgs.fromBundle(getArguments()).getImages();
        GalleryPageAdapter.setSetUIVisibility(this);
        pager.setAdapter(new GalleryPageAdapter(images));
        setUiVisibility();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        isUiVisible = false;
        setUiVisibility();
    }

    @Override
    public void setUiVisibility() {
        Window window = requireActivity().getWindow();
        View decorView = window.getDecorView();
        BottomNavigationView bottom_nav = requireActivity().findViewById(R.id.bottom_nav);
        ActionBar actionBar = ((AppCompatActivity) requireActivity()).getSupportActionBar();
        if (isUiVisible) {
            if (Build.VERSION.SDK_INT >= 30) {
                decorView.getWindowInsetsController().hide(WindowInsets.Type.navigationBars() | WindowInsets.Type.statusBars());
                decorView.getWindowInsetsController().setSystemBarsBehavior(WindowInsetsController.BEHAVIOR_SHOW_BARS_BY_SWIPE);
            } else {
                decorView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_IMMERSIVE
                        | View.SYSTEM_UI_FLAG_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION);
            }
            actionBar.hide();
            bottom_nav.setVisibility(View.GONE);
            isUiVisible = false;
        } else {
            if (Build.VERSION.SDK_INT >= 30) {
                decorView.getWindowInsetsController().show(WindowInsets.Type.statusBars() | WindowInsets.Type.navigationBars());
            } else {
                int flags = 0;
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    flags |= View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR;
                }
                decorView.setSystemUiVisibility(flags);
            }
            actionBar.show();
            bottom_nav.setVisibility(View.VISIBLE);
            isUiVisible = true;
        }
    }
}

File: app/src/main/java/pl/vemu/zsme/ui/detail/GalleryFragment.kt
package pl.vemu.zsme.ui.detail

import android.os.Build
import android.os.Bundle
import android.view.*
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.navArgs
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.bottomnavigation.BottomNavigationView
import pl.vemu.zsme.R

class GalleryFragment : Fragment() {


    private lateinit var pager: ViewPager2
    private var isUiVisible = true

    private val args: GalleryFragmentArgs by navArgs()


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        pager = ViewPager2(requireContext())
        pager.layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, MATCH_PARENT) //TODO remove?
        return pager
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        pager.adapter = GalleryPageAdapter(this, args.images)
        switchUiVisibility()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        isUiVisible = false
        switchUiVisibility()
    }

    @Suppress("DEPRECATION")
    fun switchUiVisibility() {
        val decorView = requireActivity().window.decorView
        val bottomNav: BottomNavigationView = requireActivity().findViewById(R.id.bottom_nav)
        val actionBar = (requireActivity() as AppCompatActivity).supportActionBar
        if (isUiVisible) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                decorView.windowInsetsController?.hide(WindowInsets.Type.navigationBars() or WindowInsets.Type.statusBars())
                decorView.windowInsetsController?.systemBarsBehavior = WindowInsetsController.BEHAVIOR_SHOW_BARS_BY_SWIPE
            } else {
                decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_IMMERSIVE
                        or View.SYSTEM_UI_FLAG_FULLSCREEN
                        or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION)
            }
            actionBar?.hide()
            bottomNav.visibility = View.GONE
            isUiVisible = false
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                decorView.windowInsetsController?.show(WindowInsets.Type.statusBars() or WindowInsets.Type.navigationBars())
            } else {
                decorView.systemUiVisibility = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
                    View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
                else 0
            }
            actionBar?.show()
            bottomNav.visibility = View.VISIBLE
            isUiVisible = true
        }
    }
}